src/function/scalar/system/index_key.cpp

```diff
@@ -0,0 +1,259 @@
+		const auto &struct_child_type = child_types[entry->second].second;
+		if (struct_child_type != index_types[i]) {
+			throw BinderException("index_key key_struct field '%s' has type %s but index expects %s", column_names[i],
+			                      struct_child_type.ToString().c_str(), index_types[i].ToString().c_str());
+		}
+	}
+	return field_map;
+}
+
+struct IndexKeyBindData : public FunctionData {
+	IndexKeyBindData(optional_ptr<ART> art_index_p, vector<LogicalType> index_types, vector<idx_t> field_map)
+	    : art_index(art_index_p), index_types(std::move(index_types)), field_map(std::move(field_map)) {
+	}
+
+	unique_ptr<FunctionData> Copy() const override {
+		return make_uniq<IndexKeyBindData>(*this);
+	}
+
+	bool Equals(const FunctionData &other_p) const override {
+		auto &other = other_p.Cast<IndexKeyBindData>();
+		return art_index == other.art_index && index_types == other.index_types && field_map == other.field_map;
+	}
+
+	optional_ptr<ART> art_index;
+	vector<LogicalType> index_types;
+	// Mapping from offset in key_types (a column in the index) to its corresponding offset in the struct.
+	vector<idx_t> field_map;
+};
+
+unique_ptr<FunctionData> IndexKeyBind(ClientContext &context, ScalarFunction &bound_function,
+                                      vector<unique_ptr<Expression>> &arguments) {
+	if (arguments.size() != 5) {
+		throw BinderException(
+		    "index_key requires five arguments: catalog, schema, table, index_or_constraint, key_struct");
+	}
+
+	string catalog_name = GetStringArgument(context, *arguments[0], "catalog", INVALID_CATALOG, true);
+	string schema_name = GetStringArgument(context, *arguments[1], "schema", DEFAULT_SCHEMA, true);
+	string table_name = GetStringArgument(context, *arguments[2], "table", string(), false);
+	string index_name = GetStringArgument(context, *arguments[3], "index_or_constraint", string(), false);
+
+	auto &table_entry = Catalog::GetEntry(context, CatalogType::TABLE_ENTRY, catalog_name, schema_name, table_name)
+	                        .Cast<TableCatalogEntry>();
+	auto &duck_table = table_entry.Cast<DuckTableEntry>();
+	auto &data_table = duck_table.GetStorage();
+	auto &data_table_info = *data_table.GetDataTableInfo();
+
+	data_table_info.BindIndexes(context, ART::TYPE_NAME);
+
+	auto &index_list = data_table_info.GetIndexes();
+	auto &columns = duck_table.GetColumns();
+
+	auto found_index = FindIndexByName(index_list, index_name, catalog_name, schema_name, table_entry.name);
+	ValidateIndex(*found_index, catalog_name, schema_name, table_entry.name, index_name);
+
+	auto &bound_index = found_index->Cast<BoundIndex>();
+	auto &art_index = bound_index.Cast<ART>();
+	vector<LogicalType> index_types = art_index.logical_types;
+	if (index_types.empty()) {
+		throw CatalogException("index_key: index '%s' has no key columns", index_name);
+	}
+	vector<string> column_names = GetColumnNames(bound_index, columns);
+
+	auto &key_expr = *arguments[4];
```

Comment:
I think any access to the key fields should happen in the execution function and not in the binding phase. In the execution phase we have the value for each row of the input and can correctly handle `NULL`, column references, etc.